rosidl_generator_cpp/resource/msg__struct.hpp.em

```diff
@@ -81,55 +83,140 @@ struct @(spec.base_type.type)_
 {
   using Type = @(spec.base_type.type)_<ContainerAllocator>;
 
-@# constructors (with and without allocator)
-@[for (alloc_type, alloc_name) in [['', ''], ['const ContainerAllocator & ', '_alloc']]]@
-  @('explicit ' if alloc_type else '')@(spec.base_type.type)_(@(alloc_type + alloc_name))
-@# generate initializer lists
-@[  if alloc_name == '_alloc']@
-// *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
-@[  end if]@
 @{
-used_alloc_name = False
-leading_colon = ':'
+def array_init_list(field, args):
+    tmp = [field.type.type + args] * field.type.array_size
+    return "%s{{%s}}" % (field.name, ', '.join(tmp))
```

Comment:
For a static array of size `N` this will generate the following in source code:

```
field_name{{SubmsgClass(_init), SubmsgClass(_init), ... <repeated N times>}}
```

For larger `N` that is pretty extreme. Imagine someone using `geometry_msgs::Vector3[1024] a_few_vectors`. In this case I don't think we can use the initialization list.

Reply:
Yeah, I worried about this, and kicked the can down the road.  Now we've reached it again :).  The real problem is that `std::array` element construction/initialization happens before the body of the constructor, so we can't defer it until then and make sure we pass `_init` to the constructor.  The other way I can think of to fix this is to use a `std::shared_ptr<std::array>` instead of `std::array` (and then do a `make_shared` in the constructor body), but that makes accessing fixed-size arrays different from accessing everything else.  Other ideas?

Reply:
As we discussed, the idea we are going to try here is:

1.  For fixed-size arrays with small numbers of members (say, 10), do the initialization of them in the initializer list like we are currently doing.
1.  For fixed-sized arrays that are larger than that, let the initializer run as it normally would (i.e. let it call the default constructor of the class, with no _init argument).  In the body of the constructor, we would then hand-construct objects with the appropriate _init argument, and replace the objects in the std::array with the new ones.  This has a bit of a performance impact, but should generally be acceptable.

Reply:
Thinking about this: is that worth the effort and complexity in the template? Or should we just do 2. with the argument that for such small N the double initialization won't hurt us "significantly"?